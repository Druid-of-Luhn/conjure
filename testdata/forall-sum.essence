language Essence 2.0

letting digit be domain int(0..9)

find x : digit
find y : digit

letting xy be x * y

letting myForall be quantifier {
    append  { x:bool, y:bool -> x /\ y }
    guard   { x:bool, y:bool -> x => y }
    identity true }

letting mySum be quantifier {
    append   { x:int , y:int -> x + y }
    guard    { x:bool, y:int -> x * y }
    identity 0 }

such that
    myForall i : int(1..9) . xy > i,
    (mySum i : int(1..9) . i) > xy
