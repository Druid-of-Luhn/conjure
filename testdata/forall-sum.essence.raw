Spec
  { language = "Essence"
  , version = [ 2
              , 0
              ]
  , topLevelBindings = [ ( Letting
                         , "digit"
                         , DomainIntegerFromTo (
                             Just ( ValueInteger 0 ) ) (
                             Just ( ValueInteger 9 ) )
                         )
                       , ( Find
                         , "x"
                         , Identifier "digit"
                         )
                       , ( Find
                         , "y"
                         , Identifier "digit"
                         )
                       , ( Letting
                         , "xy"
                         , GenericNode
                             Times
                             [ Identifier "x"
                             , Identifier "y"
                             ]
                         )
                       , ( Letting
                         , "forall_glueOp"
                         , DeclLambda
                             [ ( "x"
                               , TypeBoolean
                               )
                             , ( "y"
                               , TypeBoolean
                               )
                             ] (
                             GenericNode
                               And
                               [ Identifier "x"
                               , Identifier "y"
                               ] )
                         )
                       , ( Letting
                         , "forall_skipOp"
                         , DeclLambda
                             [ ( "x"
                               , TypeBoolean
                               )
                             , ( "y"
                               , TypeBoolean
                               )
                             ] (
                             GenericNode
                               Imply
                               [ Identifier "x"
                               , Identifier "y"
                               ] )
                         )
                       , ( Letting
                         , "forall"
                         , DeclQuantifier (
                             Identifier "forall_glueOp" ) (
                             Identifier "forall_skipOp" ) (
                             ValueBoolean True )
                         )
                       , ( Letting
                         , "sum_glueOp"
                         , DeclLambda
                             [ ( "x"
                               , TypeInteger
                               )
                             , ( "y"
                               , TypeInteger
                               )
                             ] (
                             GenericNode
                               Plus
                               [ Identifier "x"
                               , Identifier "y"
                               ] )
                         )
                       , ( Letting
                         , "sum_skipOp"
                         , DeclLambda
                             [ ( "x"
                               , TypeBoolean
                               )
                             , ( "y"
                               , TypeInteger
                               )
                             ] (
                             GenericNode
                               Times
                               [ Identifier "x"
                               , Identifier "y"
                               ] )
                         )
                       , ( Letting
                         , "sum"
                         , DeclQuantifier (
                             Identifier "sum_glueOp" ) (
                             Identifier "sum_skipOp" ) (
                             ValueInteger 0 )
                         )
                       ]
  , topLevelWheres = []
  , objective = Nothing
  , constraints = [ ExprQuantifier
                      { quanName = "forall"
                      , quanVar = Identifier "i"
                      , quanOver = DomainIntegerFromTo (
                                     Just ( ValueInteger 1 ) ) (
                                     Just ( ValueInteger 9 ) )
                      , quanGuard = Nothing
                      , quanBody = GenericNode
                                     Gt
                                     [ Identifier "xy"
                                     , Identifier "i"
                                     ]
                      }
                  , GenericNode
                      Gt
                      [ ExprQuantifier
                          { quanName = "sum"
                          , quanVar = Identifier "i"
                          , quanOver = DomainIntegerFromTo (
                                         Just ( ValueInteger 1 ) ) (
                                         Just ( ValueInteger 9 ) )
                          , quanGuard = Nothing
                          , quanBody = Identifier "i"
                          }
                      , Identifier "xy"
                      ]
                  ]
  , metadata = []
  }