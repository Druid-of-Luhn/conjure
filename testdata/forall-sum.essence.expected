language Essence 2.0

letting digit be domain int(0..9)
find x : digit
find y : digit
letting xy be x * y
letting forall_glueOp be lambda { x : bool, y : bool -> x /\ y }
letting forall_skipOp be lambda { x : bool, y : bool -> x => y }
letting forall be quantifier { forall_glueOp
                               forall_glueOp
                               true
                             }
letting sum_glueOp be lambda { x : int, y : int -> x + y }
letting sum_skipOp be lambda { x : bool, y : int -> x * y }
letting sum be quantifier { sum_glueOp
                            sum_glueOp
                            0
                          }

such that
    forall i : int(1..9) . xy > i,
    (sum i : int(1..9) . i) > xy
